src/util/effect.ts

```diff
@@ -0,0 +1,11 @@
+export class Effect extends Error {
```

Comment:
Why does it have to be a "class" and not just a simple function that returns an "Error" object?

Reply:
Effect is a new error type based on the [Error Cause working-proposal](https://github.com/tc39/proposal-error-cause)

Reply:
Switch to using an `addCause` function as proposed by William.

**error-cause.d.ts**
```ts
declare interface Error {
  cause: Error
}
```

**error-cause.ts**
```ts
export function addCause(cause: Error, err: Error): Error {
//...
}
```